internal/controlplane/common.go

```diff
@@ -95,3 +100,39 @@ func getProviderFromRequestOrDefault(
 	}
 	return providers[i], nil
 }
+
+func encodeListRepositoriesByProjectIDCursor(projectId, provider string, repoId int32) (string, error) {
+	if projectId == "" || provider == "" || repoId <= 0 {
+		return "", fmt.Errorf("invalid cursor elements")
+	}
+	key := strings.Join([]string{projectId, provider, strconv.Itoa(int(repoId))}, cursorDelimiter)
+	return util.EncodeCursor(key), nil
+}
+
+func decodeListRepositoriesByProjectIDCursor(cursor string) (projectId, provider string, repoId int32, err error) {
+	if cursor == "" {
+		return "", "", 0, nil
+	}
+
+	key, err := util.DecodeCursor(cursor)
+	if err != nil {
+		return "", "", 0, err
+	}
+
+	keyArr := strings.Split(key, cursorDelimiter)
+	if len(keyArr) != 3 {
+		return "", "", 0, fmt.Errorf("invalid cursor")
+	}
+
+	parsedRepoId, err := strconv.ParseInt(keyArr[2], 10, 32)
+	if err != nil {
+		return "", "", 0, err
+	}
+
+	projectId = keyArr[0]
+	provider = keyArr[1]
+
+	repoId = int32(parsedRepoId)
```

Comment:
Another option here is to make a small struct or protobuf and then serialize that.  In fact, you could:

```golang
type RepoCursor struct {
	ProjectId string
	Provider string
	RepoId int
}

func NewCursor(encoded string) (Cursor, error) {
	// ...
}

// String implements strings.Stringer
func (c *Cursor)String() string {
	if c == nil || c.ProjectId == "" || c.Provider == "" || c.RepoId == 0 {
		return ""
	}
	// Do the encode in this method, so all the encoding/decoding has to happen in this file
}
```

Note that we want to encapsulate knowledge of how a RepoCursor is serialized within these two methods, and not leak that implementation elsewhere, so we can change it later if we want.  (For example, we could compress the string and add an integrity SHA hash later if we were worried about people messing with the cursor tokens.)

Reply:
Valid points and a better pattern. Defined a struct for the cursor.